[Core] Stepper and Slider values should not be out of bound

adrianknight89 opened the pull request:
### Description of Change

```
<Stepper Value="{Binding Path=StepperValue, Mode=TwoWay}" Maximum="{Binding StepperMax}" />
```

should throw an exception if the value is greater than Maximum.

Made the same changes to `Slider` for consistency.

`slider.Value = -100;`

should throw an exception since Value is less than default Minimum. 

If custom range is required, first set Minimum and Maximum and then set Value. Constructor does clamping as usual. Clamping is also done on Value when Minimum and Maximum change. However, when directly setting Value, it should not be out of bounds.
### Bugs Fixed
- https://bugzilla.xamarin.com/show_bug.cgi?id=41960

dnfclas commented:
Hi __@adrianknight89__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by .NET Foundation and real humans are currently evaluating your PR.</p>

TTYL, DNFBOT;

StephaneDelcroix commented on Xamarin.Forms.Core/Slider.cs:
```diff
@@ -6,25 +6,25 @@ namespace Xamarin.Forms
 		{
 			var slider = (Slider)bindable;
-			slider.Value = slider.Value.Clamp(slider.Minimum, (double)value);
+			slider.Value = slider.Value.Clamp((double)value, slider.Maximum);
```
Do not shuffle stuffs for the sake of it

adrianknight89 replied:
The reason for this shuffling was to organize Stepper and Slider bindable properties in the same way. I was comparing validate and coerce functions and it was easier if the order was the same.

StephaneDelcroix commented on Xamarin.Forms.Core/Slider.cs:
```diff
@@ -65,18 +63,30 @@ public Slider(double min, double max, double val) : this()
+		/// <summary>
+		/// Gets or sets the highest possible Value for the slider.
+		/// Must be data-bound before Value to avoid Value being set to default Maximum if Value is greater than default Maximum.
+		/// </summary>
```
we don't do inline comments

adrianknight89 replied:
There are many summary comments in the source code. Just search for it in the solution and you'll get a list. Do you still want me to remove mine?

StephaneDelcroix commented on docs/Xamarin.Forms.Core/Xamarin.Forms/Slider.xml:
```diff
@@ -182,8 +182,7 @@ namespace FormsGallery
         <value>A double.</value>
-        <remarks>
-        </remarks>
+        <remarks>Must be data-bound before Value to avoid Value being set to default Maximum if Value is greater than default Maximum.</remarks>
```
we still consider this behavior as room for improvements, documenting it is a good start, but doesn't help much.

adrianknight89 replied:
What if we did something like this?

Coerce function for `Value`

```
var slider = (Slider)bindable;

if (!slider._isValueSet)
{
	if (slider.Value > slider.Maximum)
	slider.Maximum = slider.Value;

	if (slider.Value < slider.Minimum)
	slider.Minimum = slider.Value;

	slider._isValueSet = true;
}

return ((double)value).Clamp(slider.Minimum, slider.Maximum);
```

The behavioral change here would be that Minimum and Maximum values would be readjusted upon initial set of Value. In this case, the documentation changes can be removed (or edited to explain this change).

adrianknight89 commented:
@StephaneDelcroix Can you review again? I reverted my changes and started from scratch. I added a coerce function to Value to make sure it's not out of bound. Also added unit tests.

StephaneDelcroix commented on Xamarin.Forms.Core/Slider.cs:
```diff
@@ -28,16 +28,18 @@ public class Slider : View, IElementConfiguration<Slider>
 		{
 			var slider = (Slider)bindable;
 			return ((double)value).Clamp(slider.Minimum, slider.Maximum);
 		}, propertyChanged: (bindable, oldValue, newValue) =>
```
there's no point to both coerceValue and validateValue. I think core was right

adrianknight89 replied:
If you remove validate, then you'll face bug 41960. Coerce will set Value to 100 (user intends 150). If you keep validate, then the user knows to update Minimum and Maximum first before setting Value to 150.

igandrews replied:
I would think coerce is the way to go and it should just re-coerce the value after the min/max change which is how it works in WPF (which of course means you need to add a **CoerceValue** helper method on BindableObject and call it from the change callback of the related properties). Otherwise you have to know the current and new values for all the properties and set them in the correct order which if you're creating a new control from code behind is fine but it's not fine if you're doing this in XAML or binding the properties. And actually it doesn't make sense that the Minimum and Maximum of the Slider currently have both the ValidateValueCallback and the CoerceValueCallback. The ValidateValueCallback should be removed from those as well.

igandrews replied:
I would think coerce is the way to go and it should just re-coerce the value after the min/max change which is how it works in WPF (which of course means you need to add a **CoerceValue** helper method on BindableObject and call it from the change callback of the related properties). Otherwise you have to know the current and new values for all the properties and set them in the correct order which if you're creating a new control from code behind is fine but it's not fine if you're doing this in XAML or binding the properties. And actually it doesn't make sense that the Minimum and Maximum of the Slider currently have both the ValidateValueCallback and the CoerceValueCallback. The ValidateValueCallback should be removed from those as well.
